### PR TITLE
[core][compiled-graphs] Register custom serializers for InputAttributeNodes and pass their type hints to downstream nodes

### LINKED ISSUE #47251: [aDAG][Core] Feeding multiple inputs to aDAG and annotating with TorchTensor does not move tensors to GPU
### What happened + What you expected to happen

When feeding multiple tensors as input to an aDAG, the tensors do not get moved to GPU memory despite annotating with TorchTensor(). The same issue does not occur when only feeding one tensor to the aDAG.

### Versions / Dependencies

Ray==2.34

### Reproduction script

```python
import ray
import torch
import ray.dag
from ray.experimental.channel.torch_tensor_type import TorchTensorType

@ray.remote(num_gpus=1)
class Actor:
    def __init__(self) -> None:
        pass

    def test(self, tensor: torch.Tensor):
        return tensor.device

if __name__ == "__main__":
    ray.init()
    actor1, actor2 = Actor.remote(), Actor.remote()
    with ray.dag.InputNode() as dag_input:
        in1, in2 = dag_input[0], dag_input[1]

        # Dag 1
        in1 = in1.with_type_hint(TorchTensorType())
        in1 = actor1.test.bind(in1)

        # Dag 2
        in2 = in2.with_type_hint(TorchTensorType())
        in2 = actor2.test.bind(in2)

        dag = ray.dag.MultiOutputNode([in1, in2])

    adag = dag.experimental_compile()
    output = ray.get(adag.execute(torch.randn(2, 16), torch.tensor(1.0)))
    print(output)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

Issue #47251 states that the tensor will not be moved to the GPU even if users specify `TorchTensorType` on the input attribute node. The behavior is inconsistent with specifying `TorchTensorType` on the `InputNode`.

There are two reasons contributing to this inconsistent behavior:

* Reason 1: The `InputAttributeNode`'s type hint is not passed to the downstream DAG nodes.
* Reason 2: The driver process does not register the custom serializer for `InputAttributeNode`.

There are four changes in this PR:

* Resolve "Reason 1"
* Resolve "Reason 2"
* Refactor the code to register custom serializers for `InputNode`, `InputAttributeNode`, and output nodes in a single location.
* Add tests

## Related issue number

Closes #47251

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ruisearch42 — APPROVED** [review]
Looks good. Minor comments.

File: python/ray/dag/compiled_dag_node.py [hunk 1]
```diff
@@ -823,6 +823,8 @@ def __init__(
         # Preprocessing identifies the input node and output node.
         self.input_task_idx: Optional[int] = None
         self.output_task_idx: Optional[int] = None
+        # List of task indices that are input attribute nodes.
+        self.input_attr_task_idx_list: List[int] = []
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: update: https://github.com/ray-project/ray/pull/49236/commits/ebf5ea638c85013997b8d82225cdf483f7a8eff9

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 2]
```diff
@@ -118,6 +118,20 @@ def forward(self, inp):
         return torch.randn(10, 10)
 
 
+@ray.remote
+class Worker:
+    def __init__(self):
+        self.device = None
+
+    def no_op(self, tensor):
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: update: https://github.com/ray-project/ray/pull/49236/commits/ebf5ea638c85013997b8d82225cdf483f7a8eff9

File: python/ray/dag/tests/experimental/test_torch_tensor_dag.py [hunk 3]
```diff
@@ -1270,6 +1284,198 @@ def recv(self, tensor):
...
+        assert torch.equal(t2, cpu_tensor_2)
+
+        # Verify Step 2
+        device1 = ray.get(worker1.get_device.remote())
+        device2 = ray.get(worker2.get_device.remote())
+        assert device1.type == "cuda"
+        assert device2.type == "cuda"
+
+    @pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+    def test_partial_input_attr_nodes_with_tensor_type_hint(self, ray_start_regular):
```
[the remark this answers is not in the document]
Comment (kevin85421) [reply python/ray/dag/tests/experimental/test_torch_tensor_dag.py]: update: https://github.com/ray-project/ray/pull/49236/commits/ebf5ea638c85013997b8d82225cdf483f7a8eff9

**Pull request merged**